### PR TITLE
Fix #5 parallel use of decodeSingle, part of #105, restructuring

### LINKED ISSUE #5: parallel image processing breaks
https://github.com/serratus/quaggaJS/pull/319

i suspect that this also has something to do with why the tests don't always complete with the same results.

### PR DESCRIPTION
**Pull request by ericblade**

### CODE REVIEW

**Comment by ericblade** [comment]
@twsouthwick If you'd like to have a look at this, it seems to, rather surpisingly, fix parallel decode.  I expected that we'd also have to do some work in barcode_decoder and/or barcode_locator, but this is giving me correct results with small tests run many times.

I'm feeling a bit hesitant to publish this, since it could break down in a lot of ways (especially barcode_decoder and locator modules) so if you have a use case that can put it through some stress, that'd be fantastic.

File: src/quagga.js [hunk 1]
```diff
@@ -1,268 +1,40 @@
...
-    // load it to the module
-    BarcodeDecoder.registerReader(name, reader);
-    // then make sure any running instances of decoder and workers know about it
-    if (_context.decoder) {
-        _context.decoder.registerReader(name, reader);
-    }
-    QWorkers.registerReader(name, reader);
-}
+const instance = new Quagga();
+const _context = instance.context;
```
Comment (twsouthwick) [inline src/quagga.js:15]: I'd prefer to just access everything from `instance` itself - keeps it easier to reason about where the data is coming from.

**Comment by twsouthwick** [comment]
I'm building a blazor component on top of this. I'll try it out.

File: src/quagga.js (hunk 1 above)
Comment (ericblade) [reply src/quagga.js:15]: If that's the only suggestion wrt to the change as is, then I feel pretty good about the change itself being good.  I'll definitely take that into consideration.

**Comment by ericblade** [comment]
well, what i haven't done with this, is run it in LiveStream mode, which I think is much more likely where you'll find someone using start/stop.  I don't think that's going to change much -- I don't see much use for start/stop for people who are using it to decode on command, but I don't think it should be deprecated, because it is used for LiveStream.  

Typical initialization goes like

````Quagga.init({ options }, (err) => { if(!err) Quagga.start() })````

and then sometime after you'll be needing to use stop to turn it off.

... now, if there are people who are controlling the decoder manually, for some reason, rather than just using decodeSingle, then i might need to expose some way to create a new interface... because this doesn't currently allow you to create multiple Quaggas, it takes care of handling that internally.

**Comment by ericblade** [comment]
@twsouthwick and anyone else who might want to test this out .. i've just validated that LiveStream works on it after fixing a small error.  Not sure if you're wanting to use that or not, but if anyone can validate that this doesn't break anything for them, I think I'm ready to go to 1.0 with it.

**Pull request merged**